// knip/dist/plugins/vue/index.js
import { hasDependency } from '#p/util/plugin.js';
import { findWebpackDependenciesFromConfig } from '../webpack/index.js';
const title = 'Vue';
const enablers = ['vue'];
const isEnabled = ({ dependencies }) => hasDependency(dependencies, enablers);
const config = ['vue.config.{js,ts}'];
const resolveConfig = async (config, options) => {
    const { cwd, isProduction, manifest } = options;
    const deps = [];
    if (config.configureWebpack) {
        const baseConfig = {
            mode: 'development',
            entry: {},
            resolve: {},
            plugins: [],
            module: { rules: [] },
        };
        const modifiedConfig = typeof config.configureWebpack === 'function' ? config.configureWebpack(baseConfig) : config.configureWebpack;
        const { dependencies } = await findWebpackDependenciesFromConfig({
            config: modifiedConfig ?? baseConfig,
            cwd,
        });
        for (const dependency of dependencies)
            deps.push(dependency);
    }
    if (!isProduction &&
        manifest.scripts &&
        Object.values(manifest.scripts).some(script => /(?<=^|\s)vue-cli-service(\s|\s.+\s)lint(?=\s|$)/.test(script))) {
        deps.push('@vue/cli-plugin-eslint');
    }
    return deps;
};
export default {
    title,
    enablers,
    isEnabled,
    config,
    resolveConfig,
};
