// knip/dist/typescript/createHosts.js
import { EOL } from 'node:os';
import path from 'node:path';
import ts from 'typescript';
import { getCompilerExtensions } from '../compilers/index.js';
import { FOREIGN_FILE_EXTENSIONS } from '../constants.js';
import { SourceFileManager } from './SourceFileManager.js';
import { createCustomModuleResolver } from './resolveModuleNames.js';
import { createCustomSys } from './sys.js';
const libLocation = path.dirname(ts.getDefaultLibFilePath({}));
export const createHosts = ({ cwd, compilerOptions, entryPaths, compilers, isSkipLibs, useResolverCache, }) => {
    const fileManager = new SourceFileManager({ compilers, isSkipLibs });
    const compilerExtensions = getCompilerExtensions(compilers);
    const sys = createCustomSys(cwd, [...compilerExtensions, ...FOREIGN_FILE_EXTENSIONS]);
    const resolveModuleNames = createCustomModuleResolver(sys, compilerOptions, compilerExtensions, useResolverCache);
    const languageServiceHost = {
        getCompilationSettings: () => compilerOptions,
        getScriptFileNames: () => Array.from(entryPaths),
        getScriptVersion: () => '0',
        getScriptSnapshot: (fileName) => fileManager.getSnapshot(fileName),
        getCurrentDirectory: sys.getCurrentDirectory,
        getDefaultLibFileName: ts.getDefaultLibFilePath,
        readFile: sys.readFile,
        fileExists: sys.fileExists,
        resolveModuleNames,
    };
    const compilerHost = {
        writeFile: () => undefined,
        getDefaultLibLocation: () => libLocation,
        getDefaultLibFileName: languageServiceHost.getDefaultLibFileName,
        getSourceFile: (fileName) => fileManager.getSourceFile(fileName),
        getCurrentDirectory: languageServiceHost.getCurrentDirectory,
        getCanonicalFileName: (fileName) => fileName,
        useCaseSensitiveFileNames: () => true,
        getNewLine: () => EOL,
        readFile: languageServiceHost.readFile,
        fileExists: languageServiceHost.fileExists,
        resolveModuleNames: languageServiceHost.resolveModuleNames,
    };
    return { fileManager, compilerHost, resolveModuleNames, languageServiceHost };
};
