// knip/dist/binaries/resolvers/fallback.js
import parseArgs from 'minimist';
import { compact } from '../../util/array.js';
import { toBinary } from '../../util/protocols.js';
import { tryResolveFilePath, tryResolveSpecifiers } from '../util.js';
const withPositional = parsed => [parsed._[0], parsed.require].flat();
const withoutPositional = parsed => [parsed.require].flat();
const withoutRequire = () => [];
const argFilters = {
    adb: withoutRequire,
    deploy: withoutRequire,
    'babel-node': withPositional,
    esbuild: withPositional,
    execa: withPositional,
    vitest: withoutRequire,
    'vite-node': withPositional,
    zx: withPositional,
    default: withoutPositional,
};
const spawningBinaries = ['cross-env', 'retry-cli'];
export const resolve = (binary, args, { cwd, fromArgs }) => {
    const parsed = parseArgs(args, { string: ['r'], alias: { require: ['r', 'loader'] }, boolean: ['quiet', 'verbose'] });
    const bin = binary.startsWith('.') ? tryResolveFilePath(cwd, binary) : toBinary(binary);
    const filteredArgs = binary in argFilters ? argFilters[binary](parsed) : argFilters.default(parsed);
    const shiftedArgs = spawningBinaries.includes(binary) ? fromArgs(args) : [];
    return compact([bin, ...tryResolveSpecifiers(cwd, filteredArgs), ...shiftedArgs]);
};
